File: plugins/OrchidForms/src/main/java/com/eden/orchid/presentations/FormField.java
package com.eden.orchid.presentations;

import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.options.OptionsHolder;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.options.annotations.StringDefault;
import lombok.Getter;
import lombok.Setter;
import org.json.JSONObject;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public abstract class FormField implements OptionsHolder {

    @Getter protected final OrchidContext context;
    @Getter protected String key;

    @Option
    @StringDefault("text")
    @Getter @Setter protected String type;

    @Option @Setter protected String label;

    @Option @Setter protected String placeholder;

    @Option @Setter protected String[] templates;

    @Option @StringDefault("auto") @Setter protected String span;
    @Option @StringDefault("auto") @Setter protected String spanSm;

    @Inject
    public FormField(OrchidContext context) {
        this.context = context;
    }

    public void initialize(String key, JSONObject fieldData) {
        this.key = key;
        extractOptions(context, fieldData);
    }

    public List<String> getTemplates() {
        List<String> allTemplates = new ArrayList<>();
        if(!EdenUtils.isEmpty(templates)) {
            Collections.addAll(allTemplates, templates);
        }

        allTemplates.add("fields/" + key + ".peb");
        allTemplates.add("fields/" + type + ".peb");

        return allTemplates;
    }

    public String getLabel() {
        return (!EdenUtils.isEmpty(label)) ? label : key;
    }

    public String getPlaceholder() {
        return (!EdenUtils.isEmpty(placeholder)) ? placeholder : getLabel();
    }

    public String getSpan() {
        String wrapperClasses = "";

        switch(span) {
            case "right": wrapperClasses += "col-right "; break;
            default: wrapperClasses += "col "; break;
        }

        switch(span) {
            case "full": wrapperClasses += "col-lg-12 "; break;
            case "left": wrapperClasses += "col-lg-6 "; break;
            case "right": wrapperClasses += "col-lg-6 "; break;
            case "auto": wrapperClasses += "col-lg-6 "; break;
            default:
                try {
                    int colSpan = Integer.parseInt(span);
                    wrapperClasses += "col-lg-" + colSpan + " ";
                }
                catch (NumberFormatException e) {
                    wrapperClasses += "col-lg-6 ";
                }
                break;
        }
        switch(spanSm) {
            case "auto": wrapperClasses += "col-sm-12 "; break;
            default:
                try {
                    int colSpan = Integer.parseInt(spanSm);
                    wrapperClasses += "col-sm-" + colSpan + " ";
                }
                catch (NumberFormatException e) {
                    wrapperClasses += "col-sm-6 ";
                }
                break;
        }

        return wrapperClasses;
    }

    public abstract boolean acceptsType(String type);

}

File: plugins/OrchidForms/src/main/kotlin/com/eden/orchid/forms/model/FormField.kt
package com.eden.orchid.forms.model

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.options.annotations.BooleanDefault
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.options.annotations.StringDefault
import com.eden.orchid.api.server.annotations.Extensible
import com.eden.orchid.api.theme.components.ModularPageListItem
import com.eden.orchid.api.theme.pages.OrchidPage
import java.util.Collections
import javax.inject.Inject

@Extensible
@Description("A field to be included in a dynamically-generated form.", name = "Form Fields")
abstract class FormField
@Inject
constructor(
    val inputTypes: Array<String>
) : ModularPageListItem<FormFieldList, FormField> {

    lateinit var containingPage: OrchidPage

    @Option
    @Description("The key the field maps to in the resulting form")
    lateinit var key: String

    @Option("type")
    @Description("The field type.")
    lateinit var fieldType: String

    @Option("order")
    @Description(
        "Form fields are defined in maps which do not have a defined order, and so order in which fields " +
                "are rendered are not necessarily the order they are declared. Set this property to manually define the " +
                "ordering of each field."
    )
    var orderNum: Int = 0

    @Option
    @Description("The user-facing label of this field.")
    var label: String = ""
        get() {
            return if (!EdenUtils.isEmpty(field)) field else key
        }

    @Option
    @Description("The user-facing placeholder of this field.")
    var placeholder: String = ""
        get() {
            return if (!EdenUtils.isEmpty(field)) field else label
        }

    @Option
    @Description(
        "A list of templates to use for this field, which takes precedence over the normal template. The " +
                "first template in this lit that exists will be chosen, otherwise falling back to the default if the " +
                "list is empty or none of the templates are found."
    )
    lateinit var template: Array<String>

    @Option("span")
    @StringDefault("auto")
    @Description(
        "The number of columns this field should occupy on large screens, out of 12. Can also be 'left' (6" +
                "columns, ordered from the left), 'right' (6 columns, ordered from the right), 'auto' (equivalent to " +
                "'left'), or 'full' (12 columns) The default is 'auto'."
    )
    lateinit var spanVal: String

    @Option("spanSm")
    @StringDefault("auto")
    @Description(
        "The number of columns this field should occupy on small screens, out of 12. Can also be 'left', " +
                "'right', 'auto', or 'full', which are all 12 columns. The default is 'auto'."
    )
    lateinit var spanSmVal: String

    @Option
    @BooleanDefault(false)
    @Description("Whether this field is required for submission.")
    var required: Boolean = false

    override fun initialize(context: OrchidContext, containingPage: OrchidPage) {
        this.containingPage = containingPage
    }

    open fun getTemplates(): List<String> {
        val allTemplates = ArrayList<String>()
        if (!EdenUtils.isEmpty(template)) {
            Collections.addAll(allTemplates, *template)
        }

        allTemplates.add("fields/$type-$key")
        allTemplates.add("fields/$key")
        allTemplates.add("fields/$type")

        return allTemplates
    }

    fun getSpan(): String {
        var wrapperClasses = ""

        when (spanVal) {
            "right" -> wrapperClasses += "col-right "
            else -> wrapperClasses += "col "
        }

        when (spanVal) {
            "full" -> wrapperClasses += "col-lg-12 "
            "left" -> wrapperClasses += "col-lg-6 "
            "right" -> wrapperClasses += "col-lg-6 "
            "auto" -> wrapperClasses += "col-lg-6 "
            else -> try {
                val colSpan = Integer.parseInt(spanVal)
                wrapperClasses += "col-lg-$colSpan "
            } catch (e: NumberFormatException) {
                wrapperClasses += "col-lg-6 "
            }

        }
        when (spanSmVal) {
            "auto" -> wrapperClasses += "col-sm-12 "
            else -> try {
                val colSpan = Integer.parseInt(spanSmVal)
                wrapperClasses += "col-sm-$colSpan "
            } catch (e: NumberFormatException) {
                wrapperClasses += "col-sm-6 "
            }

        }

        return wrapperClasses
    }

    fun setSpan(span: String) {
        this.spanVal = span
    }

    fun setSpanSm(span: String) {
        this.spanSmVal = span
    }

    fun acceptsType(type: String): Boolean {
        return inputTypes.contains(type)
    }

    override fun setOrder(order: Int) {
        orderNum = order
    }

    override fun getOrder(): Int {
        return orderNum
    }

    override fun getType(): String {
        return fieldType
    }

    fun setType(type: String) {
        this.fieldType = type
    }

    override fun canBeUsedOnPage(
        containingPage: OrchidPage,
        modularList: FormFieldList,
        possibleItems: List<Map<String, Any>>,
        currentItems: MutableList<FormField>
    ): Boolean {
        return true
    }
}
